lib/datadog/statsd.rb

```diff
@@ -443,17 +446,28 @@ def send_stats(stat, delta, type, opts=EMPTY_OPTIONS)
 
     def send_stat(message)
       if @batch_nesting_depth > 0
+        message_bytes = message.bytesize
+        unless @buffer_bytes.zero?
+          if @buffer_bytes + 1 + message_bytes >= @max_buffer_bytes
+            flush_buffer
+          else
+            @buffer << NEW_LINE
+            @buffer_bytes += 1
+          end
+        end
+
         @buffer << message
-        flush_buffer if @buffer.length >= @max_buffer_size
+        @buffer_bytes += message_bytes
       else
         send_to_socket(message)
       end
     end
 
     def flush_buffer
-      return @buffer if @buffer.empty?
-      send_to_socket(@buffer.join(NEW_LINE))
-      @buffer = Array.new
+      return if @buffer_bytes.zero?
+      send_to_socket(@buffer)
+      @buffer = String.new
```

Comment:
not using `""` since that breaks the tests on ruby 2.3+ where we use frozen strings